modules: tfm: add kconfig to enable/disable tfm secure partitions

hakonfam opened the pull request:
Also update tfm module to pass them to the trusted-firmware-m
build system.

Depends on https://github.com/zephyrproject-rtos/trusted-firmware-m/pull/40

Signed-off-by: Håkon Øye Amundsen <haakon.amundsen@nordicsemi.no>

hakonfam committed:
modules: trusted-firmware-m: add kconfig to enable/disable partitions

Also update tfm module cmake to pass them to the trusted-firmware-m
build system.

Signed-off-by: Håkon Øye Amundsen <haakon.amundsen@nordicsemi.no>

hakonfam commented:
I'm not sure how to solve this dependency-wise, since after https://github.com/zephyrproject-rtos/trusted-firmware-m/pull/40 is merged things will be sort of broken until this is merged. (That is, all the partitions will be disabled).

tejlmand reviewed (dismissed):
~Looking good.~
~Consider to take in the suggestion on the argument list.~
Wrong PR. Intended for https://github.com/zephyrproject-rtos/trusted-firmware-m/pull/40

oyvindronningstad commented on modules/Kconfig.tfm:
```diff
@@ -94,6 +94,30 @@ config TFM_BL2_FALSE
+	default y
+
+config TFM_PARTITION_INITIAL_ATTESTATION
+	bool "Enable secure partition 'Initial Attestation'"
```
This depends on crypto right?
```suggestion
	bool "Enable secure partition 'Initial Attestation'"
	depends on TFM_PARTITION_CRYPTO
```

hakonfam replied:
I wanted to leave the dependency resolving out of this.

The reason is that it will, at best, be a clone of functionality which _should_ exist in `trusted-firmware-m` repository.
Then we are in the situation where we have to maintain this, possibly confusing the user if the dependencies stated here are out-of-sync with those it `trusted-firmware-m`.

What do you think?

oyvindronningstad replied:
That's fine, as long as the errors from the TFM build system are somewhat understandable and can be traced back to these configs.

oyvindronningstad reviewed (changes requested):
Can you find the other dependencies as well?

tejlmand commented on modules/Kconfig.tfm:
```diff
@@ -94,6 +94,30 @@ config TFM_BL2_FALSE
 
+config TFM_PARTITION_PROTECTED_STORAGE
+	bool "Enable secure partition 'Protected Storage'"
+	default y
```
Could you add some help text, describing what exactly this setting does, and that the reader can find additional info in Trusted Firmware M itself.

Maybe here would be a good place to state that dependencies between partitions are handled in TFM build system, and not Kconfig, as we have an exception here.

This goes for all the `config`s in this PR.

```suggestion
	help
	  This is what I do, and here you can find more info...
```

hakonfam replied:
Check now, not sure exactly what to write.

ioannisg reviewed (commented):
No objections to the Kconfig symbols, but i'd like to agree, also, what the default should be (whether they should be default y)

hakonfam commented:
> No objections to the Kconfig symbols, but i'd like to agree, also, what the default should be (whether they should be default y)

When deciding whether something should be enabled by default or not, I like to consider what will be the most common choice by a user. 

IMO most users will only enable a few TFM partitions. With that in mind I would suggest we default to `n`.

tejlmand reviewed (approved):
lgtm

ioannisg commented:
Needs rebase @hakonfam

hakonfam committed:
modules: trusted-firmware-m: fix incorrect check of <option> args

'IPC' and 'REGRESSION' are passed to the <option>
argument of cmake_parse_arguments, hence they are always defined,
and 'if (DEFINED)' would always return TRUE.

Use 'if' directly instead of 'if DEFINED' to check if these
options are set or not.

Ref: NCSDK-7702

Signed-off-by: Håkon Øye Amundsen <haakon.amundsen@nordicsemi.no>
